api/scpca_portal/models/project.py

```diff
@@ -1217,71 +1192,92 @@ def combine_metadata(self, updated_samples_metadata, libraries_metadata, sample_
-                    sample_csv_writer = csv.DictWriter(
-                        sample_file, fieldnames=field_names, delimiter=common.TAB
-                    )
-                    sample_csv_writer.writeheader()
 
-                    sample_libraries_metadata = (
-                        library
-                        for library in libraries_metadata_filtered_keys
-                        if library["scpca_library_id"]
-                        in sample_libraries_mapping.get(scpca_sample_id, set())
+                for sample_library_metadata in sample_libraries_metadata:
+                    sample_library_combined_metadata = (
+                        sample_library_metadata | sample_metadata_filtered_keys
                     )
+                    all_samples_combined_metadata[modality].append(sample_library_combined_metadata)
 
-                    for sample_library_metadata in sample_libraries_metadata:
-                        sample_library_combined_metadata = (
-                            sample_library_metadata | sample_metadata_filtered_keys
-                        )
-                        combined_metadata[modality].append(sample_library_combined_metadata)
+        return all_samples_combined_metadata
 
-                        sample_csv_writer.writerow(sample_library_combined_metadata)
+    def write_combined_metadata_libraries(self, combined_metadata: Dict[str, List[Dict]]) -> None:
+        """
+        Takes all combined_metadata as lists of libraries (accessed by modality),
+        and writes them to tsv files, beginning at the sample level,
+        and concluding at the project level. If a project or sample has multiple modalities
+        then multiple tsv files will be written as each modality is processed.
+        """
 
-                    if modality == Sample.Modalities.MULTIPLEXED:
-                        multiplexed_with_mapping = self.get_multiplexed_with_mapping(
-                            libraries_metadata[Sample.Modalities.MULTIPLEXED]
-                        )
-                        sample_csv_writer.writerows(
-                            self.get_multiplexed_with_combined_metadata(
-                                scpca_sample_id,
-                                multiplexed_with_mapping,
-                                sample_libraries_mapping,
-                                samples_metadata_filtered_keys,
-                                libraries_metadata_filtered_keys,
-                            )
+        # Pre-calculate mapping to be used for multiplexed samples
+        multiplexed_with_mapping = self.get_multiplexed_with_mapping(
+            combined_metadata[Sample.Modalities.MULTIPLEXED]
+        )
+
+        for modality in combined_metadata.keys():
+            if not combined_metadata[modality]:
+                continue
+
+            # Establish field names for tsv
+            key_set = set(combined_metadata[modality][0].keys())
```

Comment:
Do you think we should add a utility here `get_keys_for_dicts` or something along those lines, that would iterate through the list of dicts and add keys to a set that is returned? I'm not a fan of how we assume uniformity here.

Reply:
I see your point here. Let's discuss this.

Reply:
Added.